### DISCUSSION TITLE
Lagrange multiplier constraint

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#20648] [Q&A General] [answered]
Hello,
I am currently using `LinearNodalConstraint` to apply  <img src="https://render.githubusercontent.com/render/math?math={\color{blue}u^{L}  =  u^{R} %2B u^{1}}"> constraints.

There are two formulation types, penalty and kinematic. the kinematic implementation seem to produce the intended result, however, as the mesh gets finer, the convergence fails most of the time for a simple elastic test.

The way that these constraints are set up is by forming a constraint for each node pair+independent node.
so my question is:
1) Do you have any suggestion of why would the convergence become very slow/fails with the kinematic formulation
2) if I were to implement this constraint using LM, what are your suggestions for doing so in MOOSE. I did see `NodalEqualValueConstraint` and it seems to do this using LM. is there a way to extend this function for my purposes?

Thanks in advance!

**Comment by KhaledNabilSharafeldin** [comment, marked as answer]
I am a bit confused on how the residual is computed in `NodalEqualValueConstraint`
so basically, we get the residual for the LM variable, and define it as
https://github.com/idaholab/moose/blob/40ea18c64df1c4941c186ada524a760be091c9d0/framework/src/kernels/NodalEqualValueConstraint.C#L34-L43

now the _value holds the constrained variable value, the way it is indexed in that for loop doesn't make sense to me.
shouldnt it be  (\*_value[0])[0] - (\*_value[1])[0] since it's a nodal variable and only the first index that holds the variable values?
also, the value changes every time the the loop itterates, so not sure how does this work.

thanks again in advance!

**Reply by KhaledNabilSharafeldin** [reply]
any suggestions on applying Nodal LM constraints in 3D would be greatly appreciated, I can see it's been used extensively in the contact module. but it's more general approach than what I need it to be since my problem is more straightforward than that.

I'm currently messing with `NodalEqualValueConstraint` along with `OneDEqualValueConstraintBC`. But I am not entirely sure on the role of `OneDEqualValueConstraintBC` in the process and how to adapt it for my purposes.

thanks again!

**Reply by GiudGiud** [reply]
@lindsayad

**Reply by lindsayad** [reply]
Can you explain more what 
![image](https://user-images.githubusercontent.com/10248304/161114741-ea0d6d0c-64cb-4cec-b1d0-1573c8370ab6.png)
means? I take L to mean "left" and R to mean "right", but I don't know what "1" denotes

**Reply by KhaledNabilSharafeldin** [reply]
```
   *      E_________H
   *      /|       /|
   *    F/_______G/ |     the x node pairs correspond to the {AB,HG} edge pairs
   *     | |     |  |     the y node pairs correspond to the {AD,FG} edge pairs
   *     |A|_____|__|D    the z node pairs correspond to the {AE,CG} edge pairs
   *     |/      | /
   *    B|_______|/C
   *
   * B->1, D->2, E->4
```
<img src="https://render.githubusercontent.com/render/math?math={\color{white}u^{1}}"> would be one of the corner nodes in the mesh, and it would be defined using a `FunctionDirichletBC`.
so basically for the equation above, <img src="https://render.githubusercontent.com/render/math?math={\color{white}u^{L}}"> would correspond to one of the nodes on the face, and <img src="https://render.githubusercontent.com/render/math?math={\color{white}u^{R}}"> is the "paired" node to it.
The LHS of the equation can be treated as a secondary variable that can vary to satisfy the primary RHS.

**Reply by KhaledNabilSharafeldin** [reply]
@lindsayad let me know if I should go over more details of what I am trying to do.

**Reply by lindsayad** [reply]
What physics is this? I've never seen something like this before. `NodalEqualValueConstraint` relies on a global (scalar) variable for the Lagrange Multiplier, however, I would imagine that you want to apply your constraint for multiple node pairs, so you would want your LM variable to be a field, not a global scalar. However, I can't visualize geometrically where the Lagrange multiplier degrees of freedom should live.

**Reply by KhaledNabilSharafeldin** [reply]
I am trying to impose this type of constraint mentioned in [section 4.5.1](https://onlinelibrary.wiley.com/doi/full/10.1002/9781119176817.ecm2107) to form homogenous RVE from micro to macro scale.

if there is a way to impose this constraint using direct elimination that would be much better solution, but I imagined that LM method might be easier.

so for example here, one of the equations used is the <img src="https://render.githubusercontent.com/render/math?math={\color{white}u^{faceR}  =  u^{faceL} %2B u^{nodeD}}"> which apply to faces of the cube. by prescribing this relation the micro fluctuations in displacement should be equal to zero at the boundaries
there exists also similar equations for the corner nodes and the cube edges
example corner: <img src="https://render.githubusercontent.com/render/math?math={\color{white}u^{nodeG}  =  u^{nodeB} %2B u^{nodeD}%2B u^{nodeE}}">
example edge: <img src="https://render.githubusercontent.com/render/math?math={\color{white}u^{edgeGH}  =  u^{edgeAB} %2B u^{nodeD}%2B u^{nodeE}}">

**Reply by lindsayad** [reply]
@dschwen are you familiar with this kind of thing?

**Reply by KhaledNabilSharafeldin** [reply]
> I would imagine that you want to apply your constraint for multiple node pairs, so you would want your LM variable to be a field

I guess I did a similar thing, where i created a global scalar for each constraint equation at each node (which is a lot of variables, but I am trying to resolve this ASAP) but I couldn't figure out what to do to adapt the `OneDEqualValueConstraintBC` for my use. 
this is the input file 
https://github.com/idaholab/moose/blob/f18a0bf011d4e1a961efa1daf85e81db72565796/test/tests/mortar/1d/1d.i#L50-L66

**Reply by lindsayad** [reply]
In multiple dimensions, that `IntegratedBC` is not going to do what you want (it seems). I think what you'll want to do to continue with your methodology of multiple SCALAR lagrange multipliers is to use something similar to `NodalEqualValueConstraint` (modified to do your uL = uR + u1 physics) as the object that defines the Lagrange multiplier constraint residual, and then use `CoupledForceNodalKernel` to apply the Lagrange Multiplier to the primal variable u. `CoupledForceNodalKernel` is a `NodalKernel` which can be either block or boundary restricted. You will want to boundary restrict, and the "boundary" (nodeset) will correspond to the nodes that you are trying to constrain

**Reply by lindsayad** [reply]
`coef` will have opposite signs for uL and uR akin to what is done for `vg` in the `1d.i` input you linked to

**Reply by KhaledNabilSharafeldin** [reply]
This is great, I'll give it a shot and see how it works out! so `variable` would be the `u` variable and `v` would be the LM SCALAR variable correct?
https://github.com/idaholab/moose/blob/a0e8a1879d6a7f981f0e09d105647ab96850b049/test/tests/nodalkernels/constraint_enforcement/upper-and-lower-bound.i#L45-L73

I'm also looking at direct elimination method using `DirichletBCBase` & `NodalBC` but I will submit another discussion for that to avoid sidetracking this one.
